### ISSUE TITLE
Successive Layer Normalization in nn.Transformer

### ISSUE THREAD

**Issue by jeremyasapp** [pytorch/pytorch#24930]
In the nn.transformer.py module, the Transformer*Layer objects always have a layer norm at the very end of their forward method. However, the main Transformer object passes additional layer norms to both the TransformerEncoder and TransformerDecoder, effectively computing layer norm twice after the encoder, and after the decoder.

This seems wrong, I'm not sure there is a need for the extra layer norms inside Transformer.

Let me know if I'm missing something!

**Comment by gchanan** [comment]
@jeremyasapp when you say "wrong" do you mean non-performant, or that it would cause a problem?

@zhangguanheng66 any thoughts?

**Comment by zhangguanheng66** [comment]
There's some debate about whether to have the layer norm and the order. Overall, there are quite a few variations for the transformer model.

**Comment by jeremyasapp** [comment]
@gchanan sorry wrong was a terrible choice of word. I meant that I don't think it's consistent with the original description of the model in the "Attention is all you need" paper. I also suspect it could make the gradient computation unnecessarily more complicated?

@zhangguanheng66 I'm not quite sure I understand what you mean. Let L be the layernom function. Right now the TransformerEncoderLayer, call it E compute L(x) at the very end of its forward method. However the TransformerEncoder does L(E(x)) which means that there is an L(L(x)) being computed. Do some variants of Transformer actually do that? Do you have any intuition as to why?

**Comment by zhangguanheng66** [comment]
> @gchanan sorry wrong was a terrible choice of word. I meant that I don't think it's consistent with the original description of the model in the "Attention is all you need" paper. I also suspect it could make the gradient computation unnecessarily more complicated?
> 
> @zhangguanheng66 I realize I might not have explained the issue well, I understand there could be an ordering question, but in this particular case it seems more of an extra. Let L be the layernom function. Right now the TransformerEncoderLayer (call it E) computes L(x) at the very end of its forward method. However the TransformerEncoder does L(E(x)) which means that there is an L(L(x)) being computed: two consecutive layer norms. Do some variants of Transformer actually do that? If so do you have any intuition as to why?

The default norm at the very end of the TransformerEncoder forward function is None. https://github.com/pytorch/pytorch/blob/8ca6220509a54a93833df81635d4f662fc4a994c/torch/nn/modules/transformer.py#L146

**Comment by jeremyasapp** [comment]
It is None in the TransformerEncoder object, but if you look at the Transformer object at the very top it passes an extra LayerNorm to both the TransformerEncoder and the TransformerDecoder, so it is effectively not None

**Comment by zhangguanheng66** [comment]
In that case, the norm ops passed to the transformer_encoder is effectively none because L(L(x)) is equal to L(x) when L is is nn.LayerNorm. 

Depending on the design of the TransformerEncoderLayer and TransformerDecoderLayer, a norm ops in the end could still be helpful.

**Comment by jeremyasapp** [comment]
I see your point, you're right that L(L(x)) is effectively L(x) but in practice it's still some extra computation in the forward and backward pass? That being said, I do understand that the motivation is that it adds flexibility for Transformer to pass down a layer norm in case a custom TransformerEncoderLayer does not have a layer norm at the end. I guess I would still argue that that the default behavior of the Transformer class is going to do a bit of unnecessary computation, so maybe a norm argument in Transformer, which defaults to None, to be passed down to TransformerEncoder and TransformerDecoder would be appropriate?

**Comment by andreaskoepf** [comment]
I think @jeremyasapp is right that passing a `LayerNorm` together with a `TransformerEncoderLayer` to the `TransformerEncoder` ctor is adding a redundant norm pass. 

Since this is the code that specifically runs when no custom encoder or decoder is provided I think the `encoder_norm` and `decoder_norm` variables should be removed.

https://github.com/pytorch/pytorch/blob/c6fe864db3e17830bf12957a64e6fd579ddeffad/torch/nn/modules/transformer.py#L39-L51

**Comment by npuichigo** [comment]
Agree. Here's the encoder structure of `nn.Transformer(nhead=2, num_encoder_layers=2)`
```
(encoder): TransformerEncoder(
    (layers): ModuleList(
      (0): TransformerEncoderLayer(
        (self_attn): MultiheadAttention(
          (out_proj): _LinearWithBias(in_features=512, out_features=512, bias=True)
        )
        (linear1): Linear(in_features=512, out_features=2048, bias=True)
        (dropout): Dropout(p=0.1, inplace=False)
        (linear2): Linear(in_features=2048, out_features=512, bias=True)
        (norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (norm2): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (dropout1): Dropout(p=0.1, inplace=False)
        (dropout2): Dropout(p=0.1, inplace=False)
      )
      (1): TransformerEncoderLayer(
        (self_attn): MultiheadAttention(
          (out_proj): _LinearWithBias(in_features=512, out_features=512, bias=True)
        )
        (linear1): Linear(in_features=512, out_features=2048, bias=True)
        (dropout): Dropout(p=0.1, inplace=False)
        (linear2): Linear(in_features=2048, out_features=512, bias=True)
        (norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (norm2): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (dropout1): Dropout(p=0.1, inplace=False)
        (dropout2): Dropout(p=0.1, inplace=False)
      )
    )
    (norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
  )
```
The LayerNorm layers are duplicated.
